File: Data Structures & Algorithms/trapping-rain-water/submission-1.java
class Solution {
    public int trap(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int lMax = 0;
        int rMax = 0;
        int total = 0;

        while (l <= r) {
            if (lMax < rMax) {
                if (height[l] <= lMax) {
                    total += Math.min(lMax, rMax) - height[l];
                }
                else {
                    lMax = height[l];
                }

                l++;
            }
            else {
                if (height[r] <= rMax) {
                    total += Math.min(lMax, rMax) - height[r];
                }
                else {
                    rMax = height[r];
                }

                r--;
            }

        }

        return total;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        best = 0
        while l < r:
            water = min([height[l], height[r]]) * (r - l)
            best = max([water, best])
            if (height[l] > height[r]):
                r -= 1
            else:
                l += 1

        return best